src/viur/core/modules/user.py

```diff
@@ -827,6 +830,139 @@ def transaction(user_key, idx):
+            return self.userModule.render.secound_factor_add(
+                otp_uri=AuthenticatorOTP.generate_otp_secret_key_uri(otp_secret_key))
+        else:
+            if not securitykey.validate(skey):
+                raise errors.PreconditionFailed()
+            if not AuthenticatorOTP.verify_otp(otp, otp_secret_key):
+                return self.userModule.render.secound_factor_add(
+                    otp_uri=AuthenticatorOTP.generate_otp_secret_key_uri(otp_secret_key))  # to add errors
+
+            # Now we can set the otp_secret_key to the current User and render der Success-template
+            AuthenticatorOTP.set_otp_secret_key(otp_secret_key)
+            return self.userModule.render.secound_factor_add_success()
+
+    def canHandle(self, user_key) -> bool:
+        """
+            We can only handle the second factor if we have stored an otp_secret_key before.
+        """
+        if not (user := db.Get(user_key)):
+            return False
+
+        return len(str(user.get("otp_secret_key", ""))) > 0
+
+    @classmethod
+    def get2FactorMethodName(*args, **kwargs):
+        return "X-VIUR-2FACTOR-AuthenticatorOTP"
+
+    @classmethod
+    def set_otp_secret_key(cls, otp_secret_key=None):
+        """
+             Write a new OTP Token in the current user entry.
+        """
+
+        if otp_secret_key is None:
+            logging.error("No 'otp_secret_key' is provided")
+            raise errors.PreconditionFailed("No 'otp_secret_key' is provided")
+        if not (cuser := current.user.get()):
+            raise errors.Unauthorized()
+
+        def transaction(user_key):
+
+            if not (user := db.Get(user_key)):
+                raise errors.NotFound()
+            user["otp_secret_key"] = otp_secret_key
+            db.Put(user)
+
+        db.RunInTransaction(transaction, cuser["key"])
+
+    @classmethod
+    def generate_otp_secret_key_uri(cls, otp_secret_key):
+        """
+            :return an otp uri like otpauth://totp/Example:alice@google.com?secret=ABCDEFGH1234&issuer=Example
+        """
+
+        if not (cuser := current.user.get()):
+            raise errors.Unauthorized()
+        if not (issuer := conf["viur.otp.issuer"]):
+            logging.warning(
+                f"""conf["viur.otp.issuer"] is None we replace the issuer by conf["viur.instance.project_id"]""")
+            issuer = conf["viur.instance.project_id"]
+
+        return pyotp.TOTP(otp_secret_key).provisioning_uri(name=cuser["name"], issuer_name=issuer)
+
+    @classmethod
+    def generate_otp_secret_key(cls):
+        """
+            Generate a new OTP Secret
+            :return an otp
+        """
+        return pyotp.random_base32()
+
+    @classmethod
+    def verify_otp(cls, otp, secret):
+        return pyotp.TOTP(secret).verify(otp)
+
+    def startProcessing(self, user_key):
+        return self.userModule.render.edit(TimeBasedOTP.OtpSkel(), action="otp", tpl=self.otp_template)
```

Comment:
```suggestion
        return self.userModule.render.edit(TimeBasedOTP.OtpSkel(), action="authenticatorOTP", tpl=self.otp_template)
```

or with the new member variable above:


```suggestion
        return self.userModule.render.edit(TimeBasedOTP.OtpSkel(), action=self.ACTION_NAME, tpl=self.otp_template)
```